tests/docs/test_complete_examples.py

```diff
@@ -0,0 +1,185 @@
+                # NB: This drops the incomplete block entirely.
+                warnings.warn(
+                    f"In {example.source_filename}, a new ehrQL code block starts on line {line_number} before the previous one has been closed."
+                )
+            in_ehrql_block = True
+            example = DatasetDefinitionExample(
+                # TODO: at the moment, this displays the full absolute path.
+                # Should we have better filename handling; stripping the filename from the ehrQL path?
+                source_filename=file.name,
+                start=line_number,
+            )
+            continue
+
+        # If we are not in a ehrQL code block,
+        # and the line does not mark the start of an ehrQL code block,
+        # we can ignore the line.
+        if not in_ehrql_block:
+            continue
+
+        # This marks the end of the current ehrQL code block,
+        # so store the example and look for further examples.
+        # Whitespace following the code block markers is permitted,
+        # and we need to strip the newline to match.
+        if line.rstrip() == "```":
+            in_ehrql_block = False
+            example.end = line_number
+            yield example
+
+            example = None
+            continue
+
+        # If we are in an ehrQL code block,
+        # and the current line does not close it,
+        # append the line to the example.
+        example.content_lines.append(line)
+
+    if example is not None:
+        # TODO: Should this fail completely?
+        # NB: Incomplete examples are dropped.
+        warnings.warn(
+            f"In {example.source_filename}, an ehrQL code block starting on line {example.start} is not closed."
+        )
+
+
+def generate_complete_ehrql_examples() -> Iterator[DatasetDefinitionExample] | None:
+    """Generate all complete ehrQL examples in the documentation."""
+    # This does use a generator, although apparently pytest turns all
+    # parametrized generators into lists anyway.
+    markdown_paths = discover_markdown_paths()
+    assert len(markdown_paths) > 0, "No Markdown files found"
+
+    for p in markdown_paths:
+        with open(p) as f:
+            yield from find_complete_ehrql_examples_in_markdown(f)
+
+
+def create_example_test_case_id(example: DatasetDefinitionExample) -> str:
+    """Generate helpful test case IDs for pytest from a DatasetDefinitionExample."""
+    return f"{example.source_filename} L{example.start}:{example.end}"
+
+
+@pytest.mark.parametrize(
+    "example", generate_complete_ehrql_examples(), ids=create_example_test_case_id
+)
+# TODO: This code only handles examples that run with generate_dataset.
+# Do we need to handle measures?
+def test_ehrql_dataset_example(example: DatasetDefinitionExample) -> None:
+    """Check that an ehrQL generate dataset example works."""
+    # We need to have actual file paths to pass to generate_dataset.
```

Comment:
context: This is because the dataset definition loading uses [`importlib.util.spec_from_file_location`](https://docs.python.org/3/library/importlib.html#importlib.util.spec_from_file_location), which takes a file path and then loads it, so we can't pass in a `StringIO` or similar.

Maybe there's a better way?